### PR TITLE
Add schema for lead-options in project

### PR DESCRIPTION
**Pull request by k9845**
Addresses https://github.com/the-deep/server/issues/846

## Changes
- Added attributes 
  - lead_group
  - emm_entities
  - organization_type
  - lead_emm_trigger

Mention related users here if any.

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [x] tests
- [ ] permission checks (tests here too)
- [ ] translations

### CODE REVIEW

File: apps/project/schema.py [hunk 1]
```diff
@@ -28,6 +36,21 @@ class Meta:
...
+    def resolve_organization_type(root, info, **kwargs):
+        return OrganizationType.objects.all()
+
+    @staticmethod
+    def resolve_emm_entities(root, info, **kwargs):
+        return info.context.dl.project.emm_entities.load(root.pk)
+
+    @staticmethod
+    def resolve_lead_group(root, info, **kwargs) -> Union[int, None]:
+        return info.context.dl.project.lead_group.load(root.pk)
```
Comment (thenav56) [inline apps/project/schema.py]: Let's define this in their respective schemas.py. We also need a single object and pagination for each query.

File: apps/project/dataloaders.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class DataLoaders(WithContextMixin):
+    @cached_property
+    def emm_entities(self):
+        return EmmEntityLoader(context=self.context)
+
+    @cached_property
+    def lead_group(self):
+        return LeadGroupLoader(context=self.context)
```
Comment (thenav56) [inline apps/project/dataloaders.py]: Emm and lead_group are entities of lead, Let's define this loader in lead.

**Review by thenav56 — CHANGES_REQUESTED** [review]

**Comment by thenav56** [comment]
@k9845 Rebase required with project/graphql.

File: apps/project/schema.py [hunk 3]
```diff
@@ -32,6 +33,10 @@ class Meta:
         ), required=True
     )
 
+    @staticmethod
+    def resolve_organization_type(root, info, **kwargs):
+        return OrganizationType.objects.all()
+
```
Comment (thenav56) [inline apps/project/schema.py]: Remove this.

File: apps/project/schema.py [hunk 4]
```diff
@@ -8,6 +8,7 @@
 from utils.graphene.types import CustomDjangoListObjectType
 from utils.graphene.fields import DjangoPaginatedListObjectField
 from deep.permissions import ProjectPermissions as PP
+from organization.models import OrganizationType
```
Comment (thenav56) [inline apps/project/schema.py]: Remove this.

File: apps/organization/tests/test_schemas.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        '''
+        OrganizationType.objects.all().delete()
+        OrganizationTypeFactory.create_batch(3)
+        user = UserFactory.create()
+
+        # Without authentication -----
+        self.query_check(query, assert_for_error=True)
+
+        self.force_login(user)
+        content = self.query_check(query, assert_for_error=False)
```
Comment (thenav56) [inline apps/organization/tests/test_schemas.py]: `assert_for_error=False` is default value.

File: apps/organization/tests/test_schemas.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        self.assertEqual(len(content['data']['organizationTypes']['results']), 3, content)
```
Comment (thenav56) [inline apps/organization/tests/test_schemas.py:32]: Also assertEqual `totalCount`

File: apps/organization/schema.py [hunk 7]
```diff
@@ -72,3 +79,13 @@ class Query:
             page_size_query_param='pageSize'
         )
     )
+    organization_types = DjangoPaginatedListObjectField(
```
Comment (thenav56) [inline apps/organization/schema.py:83]: Let's add for single organization_type also.

File: apps/lead/factories.py [hunk 8]
```diff
@@ -24,9 +29,34 @@ def authors(self, create, extracted, **kwargs):
...
 class EmmEntityFactory(DjangoModelFactory):
     name = factory.Sequence(lambda n: f'emm-name-{n}')
 
     class Meta:
         model = EMMEntity
+
+
+class LeadGroupFactory(DjangoModelFactory):
+    title = factory.Sequence(lambda n: f'LeadGroup-{n}')
+    project = factory.SubFactory(ProjectFactory)
```
Comment (thenav56) [inline apps/lead/factories.py]: Let's not define SubFactory for parent models.

File: apps/lead/factories.py [hunk 9]
```diff
@@ -24,9 +29,34 @@ def authors(self, create, extracted, **kwargs):
...
+
+    class Meta:
+        model = LeadGroup
+
+
+class LeadEMMTriggerFactory(DjangoModelFactory):
+    lead = factory.SubFactory(LeadFactory)
```
Comment (thenav56) [inline apps/lead/factories.py]: Let's not define SubFactory for parent models.

File: apps/lead/schema.py [hunk 10]
```diff
@@ -30,6 +32,18 @@ def get_lead_qs(info):
...
+def get_lead_group_qs(info):
+    return LeadGroup.objects.filter(project=info.context.active_project)
+
+
+def get_emm_entities_qs(info):
+    return EMMEntity.objects.filter(lead__project=info.context.active_project).distinct()
+
+
+def get_lead_emm_entities_qs(info):
+    return LeadEMMTrigger.objects.filter(lead__project=info.context.active_project)
```
Comment (thenav56) [inline apps/lead/schema.py]: Let's add another lead__confidentiality filter as well according to Permission.

File: apps/lead/tests/test_schemas.py [hunk 11]
```diff
@@ -211,6 +217,94 @@ def test_leads_fields_query(self):
...
+        self.force_login(user)
+        # test for lead group
+        content = self.query_check(query, variables={'id': project.id})
+        self.assertEqual(content['data']['project']['leadGroups']['totalCount'], 2)
+        self.assertEqual(
+          set(result['id'] for result in content['data']['project']['leadGroups']['results']),
+          set([str(lead_group1.id), str(lead_group2.id)])
+        )
+
+        # with diffrent project
```
Comment (thenav56) [inline apps/lead/tests/test_schemas.py]:
```suggestion
        # with different project
```

**Review by thenav56 — CHANGES_REQUESTED** [review]

**Review by thenav56 — APPROVED** [review]

**Review by bimal125 — APPROVED** [review]

**Pull request merged**